81-question-mark-chrysalis/technical_specification.md

```diff
@@ -0,0 +1,233 @@
+- Config.version: Configuration identifier (always 0 in the first implementation).
+- Config.created: Datetime when this config was validated and activated.
+- Config.config: Transformation configuration data.
+
+#### `AnnotatedEMPack`
+
+Purpose: holds incoming and derived annotated EM datapacks that are to be processed or published.
+
+data structure:
+
+```python
+class AnnotatedEMPack(BaseModel):
+   id: uuid
+   model_name: str
+   original_id: str | None
+   data: DataPack
+   annotation: dict
+```
+- id: Unique identifier for the AnnotatedEMPack.
+- model_name: Unique name of the model the EMPack conforms to
+- original_id: ID of the original incoming EMPack it was derived from. None if it is an original EMPack
+- data: EMPack conforming to the model identified by `model_name`
+- annotation: Object with information from other models held by the service 
+
+
+#### Database Layer
+
+The em-transformation-service interacts with a database containing core entities. Models, workflows, and routes are loaded from a configuration YAML file and kept in fully in memory. The configuration YAML file is also stored in the database after a successful validation along with the re-derived models if there is a change. AnnotatedEMPacks are populated from incoming events (from the GHGA Study Repository) and transformation outputs. They are loaded into memory on demand. Only published data is stored persistently (using an outbox DAO). Intermediate transformed data is kept in memory as long as they are needed for running a full transformation graph corresponding to a single original data.
+
+#### Transformation Configuration
+
+The transformations are configured through the models, workflows and routes.
+
+Routes define how to transform an input model into an output model by specifying the workflow to apply and the expected output model.
+
+Routes also define a graph where models are nodes and routes are directed edges. The source nodes of this graph are the EMIMs. The graph must not contain any "diamonds", i.e. there must be at most one directed path between any two models in the graph. This also implies that the graph does not contain any cycles, i.e. is a directed acyclic graph (DAG). 
+
+We enforce this stronger unique-path property because "diamond" shapes indicate unnecessary redundancy that should be avoided, and because it eliminates any ambiguity in how data are transformed.
+
+This graph structure is used to validate the transformation configuration and to determine the processing order of schemas during model derivation and configuration changes.
+
+#### Transformation Configuration Validation
+
+Transformation configuration validation includes:
+
+1. Verify that all workflows and EMIMs (i.e., models with is_ingest = true) referenced by the routes exist in the configuration.
+2. Ensure that EMIMs do not appear as the output models of any routes.
+3. Validate schemas using the SchemaPack library and workflows using the metldata library.
```

Comment:
What's the level of validation that we can check for workflows with metldata, parsability?